Improve CLI ergonomics

mathijshenquet opened the pull request:
Closes https://github.com/epidemics/covid/issues/415 
Closes https://github.com/epidemics/covid/issues/394

@eohomegrownapps I made a PR for reference, feel free to make your own or use this one.

euanong committed:
Initial migration to click CLI system

euanong committed:
Merge branch 'master' of https://github.com/epidemics/epimodel into upstream/cli-refactoring

euanong committed:
Added workflow for web upload, made changes to web_export and web_upload

- Added workflow command group along with web upload workflow
- data-CHANNEL-v4.json is now data-v4.json; this name along with version identifier can now be specified in config.yaml
- JSON files to be uploaded are first copied into an upload directory for minification
- Pretty-printing added as an option for web-export

euanong committed:
Merge branch 'master' of https://github.com/epidemics/epimodel into upstream/cli-refactoring

euanong committed:
Code linting

Black formatting

euanong committed:
More code linting

Black formatting

euanong committed:
Added docstrings and an extra workflow command

As in title

euanong committed:
Update README.md

euanong committed:
Removed requirement for DEST CSV file in import-countermeasures to exist

euanong committed:
Merge branch 'master' of https://github.com/epidemics/epimodel into upstream/cli-refactoring

eohomegrownapps commented:
@mathijshenquet thanks for making this PR :)

Fixes https://github.com/epidemics/covid/issues/415 and https://github.com/epidemics/covid/issues/394

## Testing
I was unable to test commands involving GLEAM / gsutil fully (but there shouldn't be any reason for them to not work as before, since I didn't change any of the underlying code) -- this includes
- [ ] `export-gleam-batch`
- [ ] `generate-gleam-batch`
- [ ] `import-gleam-batch`
- [ ] `web-upload`

as well as parts of the `workflow` commands involving these:

- [ ] `prepare-gleam`
- [ ] `gleam-to-web`

I was also unable to test 
- [ ] `import-countermeasures`

as I could not find a suitable `import-countermeasures` file.

## Notes
### General
Command chaining is now possible, e.g. 
```
./do action web-export out/batch.hdf5 data/sources/estimates.csv web-upload -c main
```
To allow for command chaining, **options and flags for commands must precede arguments** (if people have strong preferences against this, command chaining can be disabled and flags can follow arguments).

Command names have been changed (underscores to hyphens):
|Command|Description|
|-|-|
|`./do action import-countermeasures`|Import one CSV file from countermeasures DB.|
|`./do action update-foretold`|Fetch data from Foretold.|
|`./do action update-johns-hopkins`|Fetch data from Johns Hopkins CSSE.|
|`./do action generate-gleam-batch`|Create batch of definitions for GLEAM.|
|`./do action export-gleam-batch`|Export batch of definitions for GLEAM.|
|`./do action import-gleam-batch`|Load batch results from GLEAM.|
|`./do action web-export`|Create data export for web.|
|`./do action web-upload`|Upload data to the configured GCS bucket.|

Workflow command macros have been added:
|Command|Description|
|-|-|
|`./do workflow prepare-gleam`|Creates and exports a batch of definitions for GLEAM (i.e. `update-foretold`, `update-johns-hopkins`, `generate-gleam-batch` and `export-gleam-batch`).|
|`./do workflow gleam-to-web`|Imports GLEAM sim results and uploads them to the web (i.e. `import-gleam-batch`, `web-export` and `web-upload`).|

I don't think I changed any of the command options and arguments, but to be safe do check them again (via `--help`) before running the commands.

### Google Storage
The uploaded `$CHANNEL` files are now stored at `gs:://static-covid/static/$VERSION/$CHANNEL/` (so uploading to production would go to `gs:://static-covid/static/v4/main`). The version number is currently hard-coded in the following lines of `config.yml` (note also that the name for the data file is no longer channel-specific):

```
gs_prefix: "gs://static-covid/static/v4/"
gs_url_prefix: "https://storage.googleapis.com/static-covid/static/v4/"
gs_datafile_name: "data-v4.json"
```

**This will require changes in the front-end.**

### web-export

The `web-export` command now copies the latest export to a directory specified in `output_latest` in `config.yml` (by default this is `out/latest`).

A `--pretty-print` flag has been added that saves the JSON files produced in pretty-printed, rather than minified, form (n.b. this applies to both the timestamped directory and the copy in `out/latest`). Using the sample HDF I was given, this increases the file size of `data-v4.json` by ~5x (from ~0.6MB to ~2.5MB), and the file size of the region JSONs by ~2x (from ~0.5MB to ~1.1MB).

The main JSON file (previously named `data-CHANNEL-v4.json` and renamed to e.g. `data-main-v4.json` on upload) has now been renamed to `data-v4.json`, as per the comments above. This filename is configurable via the `gs_datafile_name` option in `config.yml`.

### web-upload
For ease of use with `gsutil`, the `web-upload` command now copies the contents of the folder to `[directory specified by output_uploads in config.yml (by default out/latest)]/[channel to upload to]`. It then validates and minifies each of the JSON files in that folder, and uploads the folder to the URL specified by `gs_prefix` in `config.yml`. For backward compatibility, any JSON file with the old name of `data-CHANNEL-v4.json` will be renamed to `data-v4.json`.

### Miscellaneous
Click has now been added to `pyproject.toml`.

The path to `config.yml` can be specified by setting the environment variable `EPI_CONFIG`. 

Bits of `exports/epidemics_org.py` have been refactored (**possible TODO: have a cleaner way of interfacing with this code rather than passing a config object to many functions**).

## Feature / issue list
- Switched epimodel `./do` script to click options https://github.com/epidemics/covid/issues/394
- Copied latest exports to out/latest (https://github.com/epidemics/covid/issues/394)
- Made `./do web_export` output a formatted JSON which is then verified and 'minified' by `./do web_upload`; added `--pretty-print` flag to web-export (https://github.com/epidemics/covid/issues/394)
- `./do web_upload` now uploads the latest thing you generated by default )https://github.com/epidemics/covid/issues/394)
- tiny feature request: an additional terminal command that does all the steps from "export gleam results" to "web_upload" (that is, batches steps 6-8 in epimodel)
- Changed google storage filenames:
> The uploaded web for $CHANNEL files go minified to gs:://static-covid/static/$VERSION/$CHANNEL/ (so uploading to production would go to gs:://static-covid/static/v4/main), the version can be hard coded for now.
> Rename data-CHANNEL-v4.json -> data-v4.json everywhere. I never liked the the special processing of this.
> This prevents the exports from accruing in gbucket over time as it should not serve (and currently does not serve) as a backup. This way web_export export-2020-... can just create a $CHANNEL/ dir (in out/ should be fine) where it minifies the content of export-2020-... and then uploads it with gsutil.
> **This would require a change on the frontend** as well but I can do that very easily.
- Command chaining (for e.g. `./do action web-export` and `./do action web-upload`)
- Commands split into `workflow` and `actions` command groups; workflows should be easily addable / customisable.

mathijshenquet commented on do:
```diff
@@ -15,211 +15,450 @@ from epimodel.gleam import Batch, batch
+    "-c",
+    "--channel",
+    type=str,
+    default="staging",
```
Could we make this argument mandatory i.e. remove the default value?

euanong replied:
Sure (I made it optional because that was how it was defined in the original file) -- out of interest could I ask your reasoning behind doing this?

mathijshenquet replied:
1. The user should always consciously pick a channel to push to.
2. I don't see why `staging` should be privileged, if anything it should be `testing` imho.

euanong replied:
Ah, that makes sense. (For the docstring, are the only channels people might want to push to `main`, `staging` and `testing`?)

mathijshenquet replied:
The channel field is also used for 'custom sites' such as the one for Pakistan (I believe they used `balochistan`). But for the docstring putting those and maybe a mention of custom sites is ok!

hnykda replied:
I am also fine with having this mandatory.

euanong replied:
Made the change :) I've made it into an argument.

mathijshenquet commented:
Thanks @eohomegrownapps! Very nice features and a great write-up. Apart from making the `--channel` argument mandatory I approve this PR.

eohomegrownapps commented:
> Thanks @eohomegrownapps! Very nice features and a great write-up. Apart from making the `--channel` argument mandatory I approve this PR.

@mathijshenquet thanks for the feedback 😄As regards testing, should I setup a GLEAM environment (and if so, is there a guide on how to do so), or would someone else be better placed to do this?

mathijshenquet commented:
This PR https://github.com/epidemics/covid/pull/440 should be all that is needed on the frontend for this change.

mathijshenquet commented:
> As regards testing, should I setup a GLEAM environment (and if so, is there a guide on how to do so), or would someone else be better placed to do this?

I actually never ran the vackend pipeline with gleamviz so I can't really help you here. @gavento could you give some pointers/direction?

hnykda commented on do:
```diff
@@ -15,211 +15,450 @@ from epimodel.gleam import Batch, batch
+    
+    ./do action generate-gleam-batch -D 2020-04-15 -c JK default.xml estimates-2020-04-15.csv
+
+    The batch file now contains all the scenario definitions and initial populations. Note the estimate input specification may change.
```
These look rather longer - can we split it into multiple lines to fit in the black's 88 chars line length? Similar with other docstrings

euanong replied:
Wrapped (I think... is there an option for Black to check for this?)

hnykda commented on do:
```diff
@@ -15,211 +15,450 @@ from epimodel.gleam import Batch, batch
+    ctx.ensure_object(dict)
+    ctx.obj["DEBUG"] = debug
+
+    logging.basicConfig(level=logging.INFO)
+    if debug:
+        logging.root.setLevel(logging.DEBUG)
```
I would personally write it this way but it's a matter of taste
```suggestion
    logging.basicConfig(level=logging.DEBUG if debug else logging.INFO)
```

euanong replied:
Made the change :)

hnykda commented on do:
```diff
@@ -15,211 +15,450 @@ from epimodel.gleam import Batch, batch
+    # TODO: This should be somewhat more versatile
+    log.info(f"Reading estimates from CSV {country_estimates} ...")
+    est = read_csv_smart(country_estimates, ctx.obj["RDS"], levels=Level.country)
+    if start_date:
+        start_date = utils.utc_date(start_date)
     else:
         start_date = d.get_start_date()
```
also minor suggestion, feel free to ignore:

```suggestion
    start_date = utils.utc_date(start_date) if start_date else d.get_start_date()
```

euanong replied:
Made the change :)

hnykda reviewed (approved):
Looks :+1: .

I also haven't run the whole pipeline with gleam, Gavento did for the most of the time. Also maybe @jez4 ?

hnykda commented:
I think we can merge, right?

hnykda commented:
@eohomegrownapps could you try run the pipeline on your side to test this and we could merge? What are the missing pieces you need?

gavento commented on do:
```diff
@@ -15,211 +15,450 @@ from epimodel.gleam import Batch, batch
+# Actions
+@cli.group(chain=True)
+@click.pass_context
+def action(ctx):
```
I think it would be better to avoid the "action" word if it is avoidable. Can you use `chain` on `cli`?

gavento replied:
Seems possible: https://click.palletsprojects.com/en/7.x/commands/#multi-command-chaining

euanong replied:
Made the change :)

gavento commented on do:
```diff
@@ -15,211 +15,450 @@ from epimodel.gleam import Batch, batch
+    help="Path to config file (default config.yaml; alternatively set EPI_CONFIG environment variable).",
+)
+@click.pass_context
+def cli(ctx, debug, config):
```
I was hoping to get rid of global state (previously `args`) but here it is again :-)

euanong replied:
Might you be able to explain why you'd like to get rid of global state? (and is config.yaml only needed for certain functions?)

gavento commented on do:
```diff
@@ -15,211 +15,450 @@ from epimodel.gleam import Batch, batch
+    """
+    Creates and exports a batch of definitions for GLEAM. 
+
+    Runs update-johns-hopkins, update-foretold, generate-gleam-batch and export-gleam-batch.
```
I would leave out updating foretold by default
* Foretold data is not used here (rather in a colab notebook)
* You need the (non-public) key, which makes this harder to just ignore

euanong replied:
Made the change :)

gavento reviewed (commented):
I like the changes, thanks for the PR!

I would not use "action" and "workflow" if that is not technically necessary - I think it makes the commands longer, there is no added functionality within the groups, and I think the combo commands among normal ones won't confuse anyone.
(Note that the everything up to and including creating the batch file is now usually done in a colab notebook, so the composite command is less necessary. But still bood to have it functional!)

I will test it now and update the review.

eohomegrownapps commented:
Thanks for the feedback @gavento @hnykda -- I'll resolve some of these changes tomorrow (as well as attempting to test with the pipeline, if this hasn't been done already) :)

gavento commented:
@eohomegrownapps In addition to the review above:
* Is it possible to be able to pass optiona as `./do action command-a PARAM1 -O X`? Currently this now assumes that `-O ..` is option for `action` rather than for `command-a`. (`./do action command-a -O X PARAM1` works)
* Could you make unit-tests for the commands? (at least for the ones that are testable locally, but especially for generate-gleam-batch and web-export) Clip should support running scripts with some pytest fixtures, I believe.
* All the commands went through fine except for `web-upload`: I get the following error for `./do action web-upload -c ttest11`:
```
[...]
  File "/home/gavento/proj/.venv/lib/python3.7/site-packages/click/decorators.py", line 21, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "./do", line 308, in web_upload
    upload_export(dir_, c, channel=channel)
  File "/home/gavento/proj/epimodel/epimodel/exports/epidemics_org.py", line 344, in upload_export
    gs_tgt = f"{gs_prefix}/{datafile}"
NameError: name 'datafile' is not defined
```
I attach the batch file (you can use it for testing `web-export`, I used estimates `data/sources/estimates-JK-2020-04-15.csv` as `ESTIMATES`) and the exported directory for testing.
[ef-data.zip](https://github.com/epidemics/epimodel/files/4584060/ef-data.zip)

euanong committed:
Changes from code review

- Made channel option in `web-upload` mandatory (into an argument)
- Wrapped docstrings to fit the 88 character per line limit
- Changed some if statements to inline if statements
- Removed the `action` and `workflow` command groups. Every command now resides in the `cli` command group (with command chaining enabled for all commands), and workflow macros are prefixed with `workflow-`.
- Foretold no longer updated by default in `workflow-gleam-to-web`
- Bugfix for `web-upload` command

AcesoUnderGlass commented:
Suggest not using upload to main (e.g. `./do web-export out/batch.hdf5 data/sources/estimates.csv web-upload -c main`) as example code anywhere.

euanong committed:
Merge branch 'master' of https://github.com/epidemics/epimodel into upstream/cli-refactoring

euanong committed:
Update README.md

eohomegrownapps commented:
@gavento 

> Is it possible to be able to pass options as ./do action command-a PARAM1 -O X? Currently this now assumes that -O .. is option for action rather than for command-a. (./do action command-a -O X PARAM1 works)

Currently (for Click) this feature and command-chaining are mutually exclusive... it is possible for this to happen, but only if I disable command chaining. (This could be viable if there are only certain series of commands that are likely to be run one after the other often enough for command chaining to be useful, in which case I can add flags / workflow macros to facilitate these)

> Could you make unit-tests for the commands? (at least for the ones that are testable locally, but especially for generate-gleam-batch and web-export) Clip should support running scripts with some pytest fixtures, I believe.

I should be able to do this (https://click.palletsprojects.com/en/7.x/testing/) -- if so, are there any 'recommended' {estimates files, hdf files etc} or should I use the ones I was sent in the Slack? (are these OK to be open-sourced?)

> All the commands went through fine except for web-upload: I get the following error for ./do action web-upload -c ttest11:

`All the commands` was this including those which required a GLEAM pipeline set up?
`I get the following error` ah sorry -- I don't think I ever reached that point in the code locally :joy: (because I don't have gsutil installed). It should be fixed now...

eohomegrownapps commented:
@AcesoUnderGlass whereabouts is this?

AcesoUnderGlass commented:
@eohomegrownapps https://github.com/epidemics/epimodel/pull/50#issuecomment-623602757 , under general.

AcesoUnderGlass commented:
in help file:


``` 3. Import the gleamviz results into the HDF batch file, generate web
  export and export the generated folder to web (Gleamviz must be stopped
  before that.) After this succeeds, you may delete the simulations from
  gleamviz.

  ./do workflow-gleam-to-web -C ttest28
  out/batch-2020-04-16T03:54:52.910001+00:00.hdf5 data/sources/estimates-
  JK-2020-04-15.csv
```

Tried this (with available file names) and got an error:


```(epienv) Elizabeths-MBP:epimodel elvan$  ./do workflow-gleam-to-web -C ttest28 out/batch-2020-05-03T00_11_49.082698+00_00.hdf5 /Users/elvan/Downloads/v2_estimates-pak-2020-04-22.csv 
Usage: do workflow-gleam-to-web [OPTIONS] BATCH_FILE ESTIMATES CHANNEL
Try 'do workflow-gleam-to-web --help' for help.

Error: no such option: -C
(epienv) Elizabeths-MBP:epimodel elvan$ 
```

euanong committed:
Update README.md

eohomegrownapps commented:
@AcesoUnderGlass thanks; good spot 😄Evidently I hadn't been as thorough as I'd thought when updating the README...

mathijshenquet commented on epimodel/exports/epidemics_org.py:
```diff
@@ -286,22 +303,48 @@ def upload_export(dir_to_export, gs_prefix, gs_url, channel="test"):
     subprocess.run(cmd, check=True)
 
-    datafile = MAIN_DATA_FILENAME.replace("CHANNEL", channel)
     gs_tgt = f"{gs_prefix}/{datafile}"
```
Missing `datafile = MAIN_DATA_FILENAME` or similar

euanong replied:
I moved `MAIN_DATA_FILENAME` to `config.yml` and set `datafile = config["gs_datafile_name"]` in L322.

eohomegrownapps commented:
(It looks like the testing job is broken -- it can't seem to install Poetry)

AcesoUnderGlass commented:
```(epienv) Elizabeths-MBP:epimodel elvan$ ./do workflow-gleam-to-web 'out/batch-2020-05-08T00_19_12.041254+00_00.hdf5 copy' /Users/elvan/Downloads/v2_estimates-pak-2020-04-22.csv elvan
INFO:do:Importing results for 253 from GLEAM into out/batch-2020-05-08T00_19_12.041254+00_00.hdf5 copy ...
INFO:epimodel.gleam.batch:Loading results from /Users/elvan/GLEAMviz-data/sims/1588897152044.574.gvh5/results.h5 ..
INFO:epimodel.gleam.batch:Loading results from /Users/elvan/GLEAMviz-data/sims/1588897152054.574.gvh5/results.h5 ..
INFO:epimodel.gleam.batch:Loading results from /Users/elvan/GLEAMviz-data/sims/1588897152064.574.gvh5/results.h5 ..
INFO:epimodel.gleam.batch:Loading results from /Users/elvan/GLEAMviz-data/sims/1588897152073.574.gvh5/results.h5 ..
INFO:epimodel.gleam.batch:Skipped 17 regions without GleamID: {'French Southern Territories', 'Saint Helena, Ascension and Tristan da Cunha', 'Heard Island and McDonald Islands', 'Bouvet Island', 'South Georgia and the South Sandwich Islands', 'Western Sahara', 'Liechtenstein', 'Holy See', 'Tokelau', 'Antarctica', 'United States Minor Outlying Islands', 'San Marino', 'British Indian Ocean Territory', 'Pitcairn', 'Andorra', 'Saint Martin (French part)', 'Palestine, State of'}
INFO:epimodel.gleam.batch:Loaded 284144 GLEAM result rows into <Batch out/batch-2020-05-08T00_19_12.041254+00_00.hdf5 copy> (resampling 1D)
--------------------------------------------------------
data
{'john_hopkins': 'johns-hopkins.csv', 'foretold': 'foretold.csv', 'rates': 'rates.csv', 'timezones': 'timezones.csv', 'un_age_dist': 'various/age_dist_un.csv'}
--------------------------------------------------------
data
{'john_hopkins': 'johns-hopkins.csv', 'foretold': 'foretold.csv', 'rates': 'rates.csv', 'timezones': 'timezones.csv', 'un_age_dist': 'various/age_dist_un.csv'}
--------------------------------------------------------
data
{'john_hopkins': 'johns-hopkins.csv', 'foretold': 'foretold.csv', 'rates': 'rates.csv', 'timezones': 'timezones.csv', 'un_age_dist': 'various/age_dist_un.csv'}
--------------------------------------------------------
data
{'john_hopkins': 'johns-hopkins.csv', 'foretold': 'foretold.csv', 'rates': 'rates.csv', 'timezones': 'timezones.csv', 'un_age_dist': 'various/age_dist_un.csv'}
--------------------------------------------------------
data
{'john_hopkins': 'johns-hopkins.csv', 'foretold': 'foretold.csv', 'rates': 'rates.csv', 'timezones': 'timezones.csv', 'un_age_dist': 'various/age_dist_un.csv'}
--------------------------------------------------------
data
{'john_hopkins': 'johns-hopkins.csv', 'foretold': 'foretold.csv', 'rates': 'rates.csv', 'timezones': 'timezones.csv', 'un_age_dist': 'various/age_dist_un.csv'}
Traceback (most recent call last):
  File "./do", line 455, in <module>
    cli(obj={})
  File "/Users/elvan/opt/anaconda3/envs/epienv/lib/python3.7/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/Users/elvan/opt/anaconda3/envs/epienv/lib/python3.7/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/Users/elvan/opt/anaconda3/envs/epienv/lib/python3.7/site-packages/click/core.py", line 1289, in invoke
    rv.append(sub_ctx.command.invoke(sub_ctx))
  File "/Users/elvan/opt/anaconda3/envs/epienv/lib/python3.7/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/elvan/opt/anaconda3/envs/epienv/lib/python3.7/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/Users/elvan/opt/anaconda3/envs/epienv/lib/python3.7/site-packages/click/decorators.py", line 21, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "./do", line 449, in workflow_gleam_to_web
    pretty_print=pretty_print,
  File "/Users/elvan/opt/anaconda3/envs/epienv/lib/python3.7/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/Users/elvan/opt/anaconda3/envs/epienv/lib/python3.7/site-packages/click/decorators.py", line 21, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "./do", line 295, in web_export
    pretty_print,
  File "/Users/elvan/EpidemicForecasting/epimodel/epimodel/exports/epidemics_org.py", line 496, in process_export
    traces_v3 = get_extra_path(config, "traces_v3")
  File "/Users/elvan/EpidemicForecasting/epimodel/epimodel/exports/epidemics_org.py", line 454, in get_extra_path
    return Path(config["data_dir"]) / config["web_export"][name]
KeyError: 'traces_v3'
Closing remaining open files:out/batch-2020-05-08T00_19_12.041254+00_00.hdf5 copy...done
(epienv) Elizabeths-MBP:epimodel elvan$ 
```

mathijshenquet commented:
@AcesoUnderGlass Your config.yaml is missing the `traces_v3` line

mathijshenquet commented:
> (It looks like the testing job is broken -- it can't seem to install Poetry)

Super weird but this was just a fluke, it worked on rerun

euanong committed:
Removed command chaining; added upload flag to web-export

euanong committed:
Merge branch 'master' of https://github.com/epidemics/epimodel into upstream/cli-refactoring

euanong committed:
Black formatting

AcesoUnderGlass commented:
CLO looks right but the actual page isn't populated

```(epienv) Elizabeths-MBP:epimodel elvan$ ./do workflow-gleam-to-web 'out/batch-2020-05-08T00_19_12.041254+00_00.hdf5 copy' /Users/elvan/Downloads/v2_estimates-pak-2020-04-22.csv elvan2
INFO:do:Importing results for 253 from GLEAM into out/batch-2020-05-08T00_19_12.041254+00_00.hdf5 copy ...
INFO:epimodel.gleam.batch:Loading results from /Users/elvan/GLEAMviz-data/sims/1588897152044.574.gvh5/results.h5 ..
INFO:epimodel.gleam.batch:Loading results from /Users/elvan/GLEAMviz-data/sims/1588897152054.574.gvh5/results.h5 ..
INFO:epimodel.gleam.batch:Loading results from /Users/elvan/GLEAMviz-data/sims/1588897152064.574.gvh5/results.h5 ..
INFO:epimodel.gleam.batch:Loading results from /Users/elvan/GLEAMviz-data/sims/1588897152073.574.gvh5/results.h5 ..
INFO:epimodel.gleam.batch:Skipped 17 regions without GleamID: {'Saint Helena, Ascension and Tristan da Cunha', 'Holy See', 'Pitcairn', 'Antarctica', 'Tokelau', 'British Indian Ocean Territory', 'Heard Island and McDonald Islands', 'South Georgia and the South Sandwich Islands', 'United States Minor Outlying Islands', 'San Marino', 'Bouvet Island', 'Palestine, State of', 'Andorra', 'Western Sahara', 'Saint Martin (French part)', 'Liechtenstein', 'French Southern Territories'}
INFO:epimodel.gleam.batch:Loaded 284144 GLEAM result rows into <Batch out/batch-2020-05-08T00_19_12.041254+00_00.hdf5 copy> (resampling 1D)
INFO:epimodel.exports.epidemics_org:Aggregating hopkins data for ['AU-WA', 'AU-NT', 'AU-ACT', 'AU-TAS', 'AU-NSW', 'AU-VIC', 'AU-SA', 'AU-QLD'] into a single code AU
INFO:epimodel.exports.epidemics_org:Aggregating hopkins data for ['CN-HE', 'CN-ZJ', 'CN-SH', 'CN-GX', 'CN-GZ', 'CN-AH', 'CN-LN', 'CN-GD', 'CN-XJ', 'CN-HK', 'CN-JS', 'CN-NX', 'CN-XZ', 'CN-TJ', 'CN-BJ', 'CN-HL', 'CN-SN', 'CN-HN', 'CN-NM', 'CN-SX', 'CN-QH', 'CN-JL', 'CN-HI', 'CN-JX', 'CN-YN', 'CN-SC', 'CN-FJ', 'CN-CQ', 'CN-HB', 'CN-MO', 'CN-SD', 'CN-GS', 'CN-HA'] into a single code CN
INFO:epimodel.exports.epidemics_org:Aggregating hopkins data for ['CA-NB', 'CA-MB', 'CA-AB', 'CA-BC', 'CA-YT', 'CA-NL', 'CA-NT', 'CA-PE', 'CA-QC', 'CA-NS', 'CA-SK', 'CA-ON'] into a single code CA
INFO:epimodel.exports.epidemics_org:Modelled {'GD', 'MS', 'SB', 'CX', 'LY', 'W-NA', 'BZ', 'KP', 'RE', 'GW', 'SS', 'TJ', 'WF', 'TO', 'PM', 'PR', 'BW', 'VI', 'GI', 'BQ', 'GF', 'GG', 'GU', 'W-AS', 'KM', 'KY', 'W-SA', 'BM', 'TM', 'YT', 'PW', 'TV', 'SX', 'AW', 'FM', 'IM', 'SL', 'BI', 'BL', 'VG', 'TW', 'CW', 'BS', 'NF', 'AI', 'SJ', 'JE', 'MP', 'YE', 'ST', 'CC', 'CG', 'LA', 'MQ', 'CD', 'GL', 'PF', 'AS', 'MH', 'GP', 'W-AF', 'FK', 'GA', 'CV', 'KN', 'MM', 'CK', 'GM', 'VU', 'AX', 'W-OC', 'SY', 'LS', 'WS', 'MW', 'NR', 'TL', 'W-EU', 'KI', 'FO', 'NC', 'TC', 'ML', 'NU', 'DM', 'MZ'}
INFO:epimodel.exports.epidemics_org:Total data availability, number of locations: {'models': 236, 'hopkins': 239, 'foretold': 57, 'rates': 197}
INFO:epimodel.exports.epidemics_org:Export requested for 150 regions: ['AE', 'AF', 'AG', 'AL', 'AM', 'AO', 'AR', 'AT', 'AU', 'AZ', 'BA', 'BB', 'BD', 'BE', 'BF', 'BG', 'BH', 'BJ', 'BN', 'BO', 'BR', 'BT', 'BY', 'CA', 'CF', 'CH', 'CI', 'CL', 'CM', 'CN', 'CO', 'CR', 'CU', 'CY', 'CZ', 'DE', 'DJ', 'DK', 'DO', 'DZ', 'EC', 'EE', 'EG', 'ER', 'ES', 'ET', 'FI', 'FJ', 'FR', 'GB', 'GE', 'GH', 'GN', 'GQ', 'GR', 'GT', 'GY', 'HN', 'HR', 'HT', 'HU', 'ID', 'IE', 'IL', 'IN', 'IQ', 'IR', 'IS', 'IT', 'JM', 'JO', 'JP', 'KE', 'KG', 'KH', 'KR', 'KW', 'KZ', 'LB', 'LC', 'LK', 'LR', 'LT', 'LU', 'LV', 'MA', 'MC', 'MD', 'ME', 'MG', 'MK', 'MN', 'MR', 'MT', 'MU', 'MV', 'MX', 'MY', 'NA', 'NE', 'NG', 'NI', 'NL', 'NO', 'NP', 'NZ', 'OM', 'PA', 'PE', 'PG', 'PH', 'PK', 'PL', 'PT', 'PY', 'QA', 'RO', 'RS', 'RU', 'RW', 'SA', 'SC', 'SD', 'SE', 'SG', 'SI', 'SK', 'SN', 'SO', 'SR', 'SV', 'SZ', 'TD', 'TG', 'TH', 'TN', 'TR', 'TT', 'TZ', 'UA', 'UG', 'US', 'UY', 'UZ', 'VC', 'VE', 'VN', 'ZA', 'ZM', 'ZW']
INFO:epimodel.exports.epidemics_org:From exported regions (N=150): {'models': 150, 'hopkins': 150, 'foretold': 51, 'rates': 149}
WARNING:epimodel.exports.epidemics_org:Some hopkins data for CA contains empty values: {'Recovered': 0, 'Confirmed': 104, 'Deaths': 104, 'Active': 104}.
INFO:epimodel.exports.epidemics_org:Writing WebExport to out/export-2020-05-08T15-54-57.847254+00-00 ...
Writing regions: 100%|████████████████████████████████████████████████████████████████████████████████████████████| 150/150 [00:00<00:00, 190.08it/s]
INFO:epimodel.exports.epidemics_org:Exported 150 regions to out/export-2020-05-08T15-54-57.847254+00-00
INFO:epimodel.exports.epidemics_org:Copied export to out/latest
INFO:epimodel.exports.epidemics_org:Copied export to out/uploads/elvan2 for uploading
INFO:epimodel.exports.epidemics_org:Uploading data folder out/uploads/elvan2 to gs://static-covid/static/v4/elvan2 ...
Copying file://out/uploads/elvan2/extdata-SC.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-BA.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-KE.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-IQ.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-PH.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-SV.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-IE.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-SN.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-JO.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-AR.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-LB.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-GR.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-AG.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-ES.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-CR.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-NO.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-SZ.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-CF.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-LV.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-VE.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-SO.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-EG.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-BT.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-TD.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-MD.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-HN.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-BY.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-ME.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-TH.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-PE.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-KH.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-AF.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-LC.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-NZ.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-BG.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-ER.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-RO.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-ID.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-CA.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-UZ.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-MN.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-SI.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-JP.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-VN.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-AM.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-DJ.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-ZM.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-BR.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-KZ.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-SE.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-CM.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-AT.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-FJ.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-ZA.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-TN.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-NI.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-GY.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-RW.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-TZ.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-CY.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-VC.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-MV.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-CL.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-AU.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-BJ.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-MC.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-SD.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-ET.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-NE.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-FR.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-DK.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-GT.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-CU.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-BF.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-AL.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-NP.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-IN.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-KW.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-SK.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-PA.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-AO.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-AZ.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-MY.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-PT.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-BE.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-IT.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-US.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-GB.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-CZ.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-MU.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-HR.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-SR.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-CO.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-SG.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-GN.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-PY.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-BH.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-MA.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-CN.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-ZW.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-FI.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-MT.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-OM.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-DE.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-EC.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-PL.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-RU.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-LR.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-IL.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-LK.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-BD.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-TT.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-MX.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-UG.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-NG.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-LU.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-NL.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-PK.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-IR.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-HT.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-BO.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-GH.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-AE.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-JM.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-SA.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-UY.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-PG.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-CI.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-GQ.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-KG.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-QA.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-KR.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-GE.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-BB.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-MK.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-DO.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-DZ.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-TR.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-TG.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-NA.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-BN.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-MG.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/data-v4.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-UA.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-CH.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-MR.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-HU.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-EE.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-IS.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-RS.json [Content-Type=application/json]...
Copying file://out/uploads/elvan2/extdata-LT.json [Content-Type=application/json]...
| [151/151 files][  1.4 MiB/  5.6 MiB] 100% Done                                
Operation completed over 151 objects/5.6 MiB.                                    
INFO:epimodel.exports.epidemics_org:Uploading main data file to gs://static-covid/static/v4/data-v4.json ...
Copying file://out/uploads/elvan2/data-v4.json [Content-Type=application/json]...
/ [1/1 files][ 90.2 KiB/512.0 KiB] 100% Done                                    
Operation completed over 1 objects/512.0 KiB.                                    
INFO:epimodel.exports.epidemics_org:File URL: https://storage.googleapis.com/static-covid/static/v4/data-v4.json
INFO:epimodel.exports.epidemics_org:Custom web URL: http://epidemicforecasting.org/?channel=elvan2
Closing remaining open files:out/batch-2020-05-08T00_19_12.041254+00_00.hdf5 copy...doneout/batch-2020-05-08T00_19_12.041254+00_00.hdf5 copy...done
(epienv) Elizabeths-MBP:epimodel elvan$ 
```

http://epidemicforecasting.org/?channel=elvan2

mathijshenquet committed:
fix relative extdata

euanong committed:
Avoid creating extraneous data-v4.json file in parent directory of web upload

AcesoUnderGlass commented:
now getting a key error when I try to import:

```(epienv) Elizabeths-MBP:epimodel elvan$  ./do import-gleam-batch 'out/batch-start.hdf5 copy' 
INFO:do:Importing results for 253 from GLEAM into out/batch-start.hdf5 copy ...
INFO:epimodel.gleam.batch:Loading results from /Users/elvan/GLEAMviz-data/sims/1588711703116.574.gvh5/results.h5 ..
INFO:epimodel.gleam.batch:Loading results from /Users/elvan/GLEAMviz-data/sims/1588711703127.574.gvh5/results.h5 ..
INFO:epimodel.gleam.batch:Loading results from /Users/elvan/GLEAMviz-data/sims/1588711703137.574.gvh5/results.h5 ..
INFO:epimodel.gleam.batch:Loading results from /Users/elvan/GLEAMviz-data/sims/1588711703145.574.gvh5/results.h5 ..
INFO:epimodel.gleam.batch:Skipped 17 regions without GleamID: {'Holy See', 'Saint Martin (French part)', 'Liechtenstein', 'Antarctica', 'South Georgia and the South Sandwich Islands', 'San Marino', 'Pitcairn', 'United States Minor Outlying Islands', 'Western Sahara', 'Palestine, State of', 'Heard Island and McDonald Islands', 'French Southern Territories', 'Andorra', 'British Indian Ocean Territory', 'Bouvet Island', 'Tokelau', 'Saint Helena, Ascension and Tristan da Cunha'}
Traceback (most recent call last):
  File "/Users/elvan/opt/anaconda3/envs/epienv/lib/python3.7/site-packages/pandas/tseries/frequencies.py", line 222, in _get_offset
    klass = prefix_mapping[split[0]]
KeyError: 'DBATCH'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/elvan/opt/anaconda3/envs/epienv/lib/python3.7/site-packages/pandas/tseries/frequencies.py", line 170, in to_offset
    offset = _get_offset(name)
  File "/Users/elvan/opt/anaconda3/envs/epienv/lib/python3.7/site-packages/pandas/tseries/frequencies.py", line 228, in _get_offset
    raise ValueError(libfreqs.INVALID_FREQ_ERR_MSG.format(name))
ValueError: Invalid frequency: DBATCH

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "./do", line 474, in <module>
    cli(obj={})
  File "/Users/elvan/opt/anaconda3/envs/epienv/lib/python3.7/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/Users/elvan/opt/anaconda3/envs/epienv/lib/python3.7/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/Users/elvan/opt/anaconda3/envs/epienv/lib/python3.7/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/elvan/opt/anaconda3/envs/epienv/lib/python3.7/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/elvan/opt/anaconda3/envs/epienv/lib/python3.7/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/Users/elvan/opt/anaconda3/envs/epienv/lib/python3.7/site-packages/click/decorators.py", line 21, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "./do", line 199, in import_gleam_batch
    info_level=logging.INFO,
  File "/Users/elvan/EpidemicForecasting/epimodel/epimodel/gleam/batch.py", line 222, in import_results_from_gleam
    pd.Grouper(freq=resample, level=2),
  File "/Users/elvan/opt/anaconda3/envs/epienv/lib/python3.7/site-packages/pandas/core/resample.py", line 1338, in __init__
    freq = to_offset(freq)
  File "/Users/elvan/opt/anaconda3/envs/epienv/lib/python3.7/site-packages/pandas/tseries/frequencies.py", line 177, in to_offset
    raise ValueError(libfreqs.INVALID_FREQ_ERR_MSG.format(freq))
ValueError: Invalid frequency: 1Dbatch file
Closing remaining open files:out/batch-start.hdf5 copy...done```

mathijshenquet commented:
@eohomegrownapps I would like to merge this sooner rather than later. Would it be possible to push back all broken tests to a separate PR?

eohomegrownapps commented:
@mathijshenquet sure -- if you cherry-pick all commits up to and including https://github.com/epidemics/epimodel/pull/50/commits/56ba463b35a04d54892797675f53887824d259d8 that should be everything apart from the unit tests.

eohomegrownapps commented:
@mathijshenquet unit tests are now in cli-unit-tests (https://github.com/epidemics/epimodel/tree/cli-unit-tests) and should be done soon :)

eohomegrownapps commented:
Update: Unit tests (https://github.com/epidemics/epimodel/tree/cli-unit-tests) now work.

Functions not currently unit-tested:
- web-upload (there's little to test unless pytest has write access to google storage)
- workflow-gleam-to-web (this is effectively a wrapper function for functions already tested -- and testing these would double the testing time)
- workflow-prepare-gleam (as above)
- import-countermeasures (this function seems to be deprecated)
- update-foretold (requires API keys)

Issues to resolve:
- I need to implement a way for the validation files that the unit tests check against to be generated (e.g. if the specification is updated)
- I found myself copying quite a lot of files over from the epimodel-covid-data repository (in fact, tests/data/cli-dir/data is a subset of epimodel/data -- n.b. I cannot just read data files from epimodel/data because the validation files would need to be regenerated every time the data files are updated) -- a nice solution for which data files to use might be to use the sample files that @AcesoUnderGlass is thinking about adding

hnykda commented:
Finally, thanks!

mathijshenquet commented:
Unfortunately this came up https://epidemicforecasting.slack.com/archives/C0100EK7N8N/p1589149666381300?thread_ts=1588935374.335200&cid=C0100EK7N8N

hnykda commented:
That's quite sad, but it means we have to revert the revert, right? :grin: 

I will try to find out how to prevent this from happening.
